http-server/src/main/java/io/airlift/http/server/HttpServerChannelListener.java

```diff
@@ -41,48 +45,55 @@ public HttpServerChannelListener(DelimitedRequestLog logger)
     }
 
     @Override
-    public void onRequestBegin(Request request)
+    protected void onBeforeHandling(Request request)
     {
         request.setAttribute(REQUEST_BEGIN_ATTRIBUTE, System.nanoTime());
     }
 
-    @SuppressWarnings("unchecked")
     @Override
-    public void onBeforeDispatch(Request request)
+    protected void onRequestRead(Request request, Content.Chunk chunk)
     {
         long requestBeginTime = (Long) request.getAttribute(REQUEST_BEGIN_ATTRIBUTE);
         request.setAttribute(REQUEST_BEGIN_TO_DISPATCH_ATTRIBUTE, System.nanoTime() - requestBeginTime);
     }
 
-    @SuppressWarnings("unchecked")
     @Override
-    public void onRequestEnd(Request request)
+    protected void onAfterHandling(Request request, boolean handled, Throwable failure)
     {
         long requestBeginTime = (Long) request.getAttribute(REQUEST_BEGIN_ATTRIBUTE);
         request.setAttribute(REQUEST_BEGIN_TO_END_ATTRIBUTE, System.nanoTime() - requestBeginTime);
     }
 
     @Override
-    public void onResponseBegin(Request request)
+    protected void onResponseBegin(Request request, int status, HttpFields headers)
     {
         if (request.getAttribute(REQUEST_BEGIN_TO_END_ATTRIBUTE) == null) {
-            onRequestEnd(request);
+            onAfterHandling(request, true, null);
         }
         request.setAttribute(RESPONSE_CONTENT_TIMESTAMPS_ATTRIBUTE, new ArrayList<Long>());
     }
 
-    @SuppressWarnings("unchecked")
     @Override
-    public void onResponseContent(Request request, ByteBuffer content)
+    protected void onResponseWrite(Request request, boolean last, ByteBuffer content)
     {
         List<Long> contentTimestamps = (List<Long>) request.getAttribute(RESPONSE_CONTENT_TIMESTAMPS_ATTRIBUTE);
         contentTimestamps.add(System.nanoTime());
     }
 
-    @SuppressWarnings("unchecked")
     @Override
-    public void onComplete(Request request)
+    public boolean handle(Request request, Response response, Callback callback)
+            throws Exception
     {
+        Exception exception = null;
+        boolean handled = false;
+
+        try {
+            handled = super.handle(request, response, callback);
+        }
+        catch (Exception e) {
+            exception = e;
```

Comment:
This exception will arrive via `onAfterHandling(Request request, boolean handled, Throwable failure)`

Reply:
EventsHandler does not expose the response in any of its callback methods, as the response should not be modified by those events, and everything readable from the response should be exposed by the different `on*` methods.

So I'm wondering why you need the response object itself.

Reply:
> I need to access both request and response in order to provide an event that the request has completed. What's the best way to achieve that?

When both the Request and the Response are done, that event is `EventsHandler.onComplete(Request, Throwable)`

Reply:
I've switched to RequestLog usage.